### PR TITLE
Added support for opentelemetry

### PR DESCRIPTION
**Pull request by OwenCorrigan76**
Incorporates Opentelemetry alongside OpenTracing to read traces on Producer, Consumer and Streams.

Added OpenTelemetry dependencies as instructed here: [1]

Required Environmental Variables:
``` 
TRACING_SYSTEM - String for activating tracing. accepted values are `jaeger`,  `opentelemetry`.
```

[1]  https://github.com/ppatierno/kafka-opentelemetry

### CODE REVIEW

File: java/kafka/streams/pom.xml [hunk 1]
```diff
@@ -35,10 +35,36 @@
             <groupId>io.jaegertracing</groupId>
             <artifactId>jaeger-client</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.opentracing.contrib</groupId>
+            <artifactId>opentracing-kafka-client</artifactId>
+        </dependency>
```
Comment (kyguy) [inline java/kafka/streams/pom.xml]: This can be removed

**Review by kyguy — APPROVED** [review]
Looks great! Just one nit!

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 2]
```diff
@@ -22,6 +21,22 @@
...
+            switch (value) {
+                case "jaeger":
+                    return KafkaConsumerExample.TracingSystem.JAEGER;
+                case "opentelemetry":
+                    return KafkaConsumerExample.TracingSystem.OPENTELEMETRY;
+                default:
+                    return null;
+            }
+        }
+    }
```
Comment (scholzj) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: TBH, the way you use it now, I don't think the Enum really adds any value. I think you should either use it right from the beginning in the config class. Or you should just have two constants defining the two types and use `equals` to compare it with the value form config. I think this applies both here and in the other examples.

If you decide to keep it and use it from the config on, it might make sense to add maybe some `NONE` tracing to indicate that it is disabled instead of passing null around.

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 3]
```diff
@@ -30,11 +45,21 @@ public static void main(String[] args) {
...
+        if (tracingSystem != null) {
+            if (tracingSystem == KafkaConsumerExample.TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
 
-            props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());
+            } else if (tracingSystem == KafkaConsumerExample.TracingSystem.OPENTELEMETRY) {
+
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentelemetry.instrumentation.kafkaclients.TracingConsumerInterceptor.class.getName());
```
Comment (scholzj) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: When I compare it with @ppatierno and his PR in Strimzi operators ... he seems to do following in addition to configuring the interceptors:

```java
            System.setProperty("otel.metrics.exporter", "none"); // disable metrics
            AutoConfiguredOpenTelemetrySdk.initialize();
```

Is this needed here as well?

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java (hunk 2 above)
Comment (mimaison) [reply java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: If you keep the enum you can also do something like in https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/ClientDnsLookup.java#L21-L39 and rely on `valueOf()`

File: java/kafka/consumer/src/main/java/KafkaConsumerConfig.java [hunk 4]
```diff
@@ -34,12 +34,14 @@ public class KafkaConsumerConfig {
     private final String oauthTokenEndpointUri;
     private final String additionalConfig;
     private final String saslLoginCallbackClass = "io.strimzi.kafka.oauth.client.JaasClientOauthLoginCallbackHandler";
+    private final String tracingSystem;
+    private static final String DEFAULT_TRACING_SYSTEM = null;
```
Comment (mimaison) [inline java/kafka/consumer/src/main/java/KafkaConsumerConfig.java]: Strings are null by default so you can remove `= null`

File: java/kafka/consumer/src/main/java/KafkaConsumerConfig.java [hunk 5]
```diff
@@ -89,14 +93,14 @@ public static Properties createProperties(KafkaConsumerConfig config) {
         props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringDeserializer");
         props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringDeserializer");
 
-        if (config.getSslTruststoreCertificates() != null)   {
+        if (config.getSslTruststoreCertificates() != null)  {
```
Comment (mimaison) [inline java/kafka/consumer/src/main/java/KafkaConsumerConfig.java]: Nit: If we're removing extra spaces, let's remove them all! Otherwise undo this change.
Same below

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java (hunk 3 above)
Comment (ppatierno) [reply java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: Jakub is right. You are not initializing the OpenTelemetry tracing at all through the SDK.
I remember that this code was in your draft PR on your repo, or? Was it lost somehow?
Setting the metrics exporter to "none" is also needed as the current status of OpenTelemetry otherwise it tries to use the "otlp" implementation for metrics exporting which we don't want at all right now.

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 6]
```diff
@@ -30,11 +45,21 @@ public static void main(String[] args) {
         Properties props = KafkaConsumerConfig.createProperties(config);
         int receivedMsgs = 0;
 
-        if (System.getenv("JAEGER_SERVICE_NAME") != null)   {
-            Tracer tracer = Configuration.fromEnv().getTracer();
-            GlobalTracer.registerIfAbsent(tracer);
+        TracingSystem tracingSystem;
+        tracingSystem = TracingSystem.forValue(config.getTracingSystem());
```
Comment (mimaison) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: Can we merge these 2 lines?

File: java/kafka/pom.xml [hunk 7]
```diff
@@ -20,6 +20,8 @@
         <slf4j-simple.version>1.6.2</slf4j-simple.version>
         <kafka.version>2.8.0</kafka.version>
         <opentracing-kafka.version>0.1.15</opentracing-kafka.version>
+        <opentelemetry-alpha.version>1.17.0-alpha</opentelemetry-alpha.version>
+        <opentelemetry.version>1.17.0</opentelemetry.version>
```
Comment (ppatierno) [inline java/kafka/pom.xml]: `1.18.0` is out, let's use that.

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 8]
```diff
@@ -30,11 +45,21 @@ public static void main(String[] args) {
...
+
+        if (tracingSystem != null) {
+            if (tracingSystem == KafkaConsumerExample.TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
 
-            props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());
+            } else if (tracingSystem == KafkaConsumerExample.TracingSystem.OPENTELEMETRY) {
```
Comment (mimaison) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: I think you should be able to drop the `KafkaConsumerExample.` prefix here

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 9]
```diff
@@ -62,4 +87,4 @@ public static void main(String[] args) {
         }
         log.info("Received {} messages", receivedMsgs);
     }
-}
+}
```
Comment (mimaison) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: I think we can keep the newline. Same in a few other files below

File: java/kafka/producer/src/main/java/KafkaProducerExample.java [hunk 10]
```diff
@@ -34,11 +48,20 @@ public static void main(String[] args) throws InterruptedException {
...
 
-            props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingProducerInterceptor.class.getName());
+            if (tracingSystem == TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
+
+                props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingProducerInterceptor.class.getName());
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
+
+                props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentelemetry.instrumentation.kafkaclients.TracingProducerInterceptor.class.getName());
```
Comment (ppatierno) [inline java/kafka/producer/src/main/java/KafkaProducerExample.java:42]: missing SDK initialization as raised by @scholzj before

File: java/kafka/streams/src/main/java/KafkaStreamsExample.java [hunk 11]
```diff
@@ -40,16 +58,28 @@ public static void main(String[] args) {
...
+
+                KafkaClientSupplier supplier = new TracingKafkaClientSupplier(tracer);
+                streams = new KafkaStreams(builder.build(), props, supplier);
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
 
-            KafkaClientSupplier supplier = new TracingKafkaClientSupplier(tracer);
-            streams = new KafkaStreams(builder.build(), props, supplier);
+                props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentelemetry.instrumentation.kafkaclients.TracingProducerInterceptor.class.getName());
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentelemetry.instrumentation.kafkaclients.TracingConsumerInterceptor.class.getName());
+                streams = new KafkaStreams(builder.build(), props);
```
Comment (ppatierno) [inline java/kafka/streams/src/main/java/KafkaStreamsExample.java]: still missing SDK initialization.

File: java/kafka/deployment-tracing.yaml [hunk 12]
```diff
@@ -95,6 +97,8 @@ spec:
               value: const
             - name: JAEGER_SAMPLER_PARAM
               value: "1"
+            - name: TRACING_SYSTEM
+              value: "jaeger"
```
Comment (mimaison) [inline java/kafka/deployment-tracing.yaml]: Do we need the quotes?

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 13]
```diff
@@ -30,11 +45,21 @@ public static void main(String[] args) {
...
+        TracingSystem tracingSystem;
+        tracingSystem = TracingSystem.forValue(config.getTracingSystem());
+
+        if (tracingSystem != null) {
+            if (tracingSystem == KafkaConsumerExample.TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
 
-            props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());
```
Comment (mimaison) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java:34]: We should keep the import and use the short name instead of this.

File: java/kafka/producer/pom.xml [hunk 14]
```diff
@@ -32,6 +32,26 @@
             <groupId>io.jaegertracing</groupId>
             <artifactId>jaeger-client</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.opentracing.contrib</groupId>
+            <artifactId>opentracing-kafka-streams</artifactId>
+        </dependency>
```
Comment (ppatierno) [inline java/kafka/producer/pom.xml:43]: why this streams related dependency in the producer application?

**Review by ppatierno — CHANGES_REQUESTED** [review]

File: java/kafka/producer/src/main/java/KafkaProducerExample.java [hunk 15]
```diff
@@ -26,6 +25,21 @@
 public class KafkaProducerExample {
     private static final Logger log = LogManager.getLogger(KafkaProducerExample.class);
 
+    public enum TracingSystem {
```
Comment (mimaison) [inline java/kafka/producer/src/main/java/KafkaProducerExample.java]: Instead of duplicating this logic, can we extract it into its own class?

File: java/kafka/streams/pom.xml (hunk 1 above)
Comment (OwenCorrigan76) [reply java/kafka/streams/pom.xml]: Removed

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java (hunk 3 above)
Comment (OwenCorrigan76) [reply java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: Above lines for initialising the SDK now added.

File: java/kafka/consumer/src/main/java/KafkaConsumerConfig.java (hunk 4 above)
Comment (OwenCorrigan76) [reply java/kafka/consumer/src/main/java/KafkaConsumerConfig.java]: Without initialising to null, I'm getting the following error - `Variable 'DEFAULT_TRACING_SYSTEM' might not have been initialized`

File: java/kafka/consumer/src/main/java/KafkaConsumerConfig.java (hunk 5 above)
Comment (OwenCorrigan76) [reply java/kafka/consumer/src/main/java/KafkaConsumerConfig.java]: Undone

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java (hunk 6 above)
Comment (OwenCorrigan76) [reply java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: Merged

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java (hunk 8 above)
Comment (OwenCorrigan76) [reply java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: Dropped without any issues.

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java (hunk 9 above)
Comment (OwenCorrigan76) [reply java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: Do you mean an extra empty line at the bottom? If so, added.

File: java/kafka/deployment-tracing.yaml (hunk 12 above)
Comment (OwenCorrigan76) [reply java/kafka/deployment-tracing.yaml]: Quotes removed

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java (hunk 13 above)
Comment (OwenCorrigan76) [reply java/kafka/consumer/src/main/java/KafkaConsumerExample.java:34]: The reason the import prefix is added here instead of keeping the import is: Jaeger (opentracing) and opentelemetry use the same class name and the only way i could differencieate between the two is to add the import in the lines of code. Is there another way to do this? Thanks. See below:
` props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG,io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());`
`props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG,io.opentelemetry.instrumentation.kafkaclients.TracingConsumerInterceptor.class.getName());`

File: java/kafka/pom.xml (hunk 7 above)
Comment (OwenCorrigan76) [reply java/kafka/pom.xml]: Bumped to 1.18.0

File: java/kafka/producer/src/main/java/KafkaProducerExample.java (hunk 10 above)
Comment (OwenCorrigan76) [reply java/kafka/producer/src/main/java/KafkaProducerExample.java:42]: Refactored to include lines for initialising the SDK.

File: java/kafka/streams/src/main/java/KafkaStreamsExample.java (hunk 11 above)
Comment (OwenCorrigan76) [reply java/kafka/streams/src/main/java/KafkaStreamsExample.java]: Refactored to below:
```props.put(StreamsConfig.CONSUMER_PREFIX + ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
   props.put(StreamsConfig.CONSUMER_PREFIX + ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
   props.put(StreamsConfig.PRODUCER_PREFIX + ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingProducerInterceptor.class.getName());
   System.setProperty("otel.metrics.exporter", "none"); // disable metrics
   AutoConfiguredOpenTelemetrySdk.initialize();
   streams = new KafkaStreams(builder.build(), props);

File: java/kafka/producer/pom.xml (hunk 14 above)
Comment (OwenCorrigan76) [reply java/kafka/producer/pom.xml:43]: Removed

File: java/kafka/consumer/pom.xml [hunk 16]
```diff
@@ -39,6 +39,22 @@
...
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-exporter-jaeger</artifactId>
+            <version>${opentelemetry.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry.instrumentation</groupId>
+            <artifactId>opentelemetry-kafka-clients-2.6</artifactId>
+            <version>${opentelemetry-alpha.version}</version>
+            <scope>compile</scope>
```
Comment (ppatierno) [inline java/kafka/consumer/pom.xml]: why it's just at compile scope? does the application work at runtime?

Comment (ppatierno) [reply java/kafka/consumer/pom.xml]: Never mind, I read "compile" but thought "provided" ... so you can anyway remove this line because "compile" is the default scope.

File: java/kafka/common/src/main/java/TracingSystem.java [hunk 17]
```diff
@@ -0,0 +1,16 @@
+public enum TracingSystem {
```
Comment (ppatierno) [inline java/kafka/common/src/main/java/TracingSystem.java:6]: this is missing the Strimzi copyright header

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 18]
```diff
@@ -30,36 +29,44 @@ public static void main(String[] args) {
...
+        if (tracingSystem != TracingSystem.NONE) {
+            if (tracingSystem == TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
 
-            props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
-        }
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
+
```
Comment (ppatierno) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java:37]: delete blank line

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 19]
```diff
@@ -30,36 +29,44 @@ public static void main(String[] args) {
...
-            for (ConsumerRecord<String, String> record : records) {
-                log.info("Received message:");
-                log.info("\tpartition: {}", record.partition());
-                log.info("\toffset: {}", record.offset());
-                log.info("\tvalue: {}", record.value());
-                if (record.headers() != null) {
-                    log.info("\theaders: ");
-                    for (Header header : record.headers()) {
-                        log.info("\t\tkey: {}, value: {}", header.key(), new String(header.value()));
+            while (receivedMsgs < config.getMessageCount()) {
```
Comment (ppatierno) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: AFAICS the consumer application won't receive messages if tracing is disabled. You have put the logic to receive within the ` if (tracingSystem != TracingSystem.NONE)`. Can you check please? It should not work anymore with tracing disabled.

File: java/kafka/producer/src/main/java/KafkaProducerExample.java [hunk 20]
```diff
@@ -34,11 +33,20 @@ public static void main(String[] args) throws InterruptedException {
...
+        if (tracingSystem != TracingSystem.NONE) {
 
-            props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingProducerInterceptor.class.getName());
+            if (tracingSystem == TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
+
+                props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingProducerInterceptor.class.getName());
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
+
```
Comment (ppatierno) [inline java/kafka/producer/src/main/java/KafkaProducerExample.java:41]: remove blank line

File: java/kafka/streams/src/main/java/KafkaStreamsExample.java [hunk 21]
```diff
@@ -40,12 +44,28 @@ public static void main(String[] args) {
...
-            KafkaClientSupplier supplier = new TracingKafkaClientSupplier(tracer);
-            streams = new KafkaStreams(builder.build(), props, supplier);
+            if (tracingSystem == TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
+
+                KafkaClientSupplier supplier = new  io.opentracing.contrib.kafka.streams.TracingKafkaClientSupplier(tracer);
+                streams = new KafkaStreams(builder.build(), props, supplier);
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
+
```
Comment (ppatierno) [inline java/kafka/streams/src/main/java/KafkaStreamsExample.java]: remove blank line

File: java/kafka/streams/src/main/java/TracingKafkaClientSupplier.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
+import io.opentelemetry.api.GlobalOpenTelemetry;
```
Comment (ppatierno) [inline java/kafka/streams/src/main/java/TracingKafkaClientSupplier.java:6]: it's missing the Strimzi copyright header

**Review by ppatierno — CHANGES_REQUESTED** [review]

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 23]
```diff
@@ -30,36 +29,43 @@ public static void main(String[] args) {
...
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentelemetry.instrumentation.kafkaclients.TracingConsumerInterceptor.class.getName());
+            } else {
+                log.error("Error: TRACING_SYSTEM {} is not recognized or supported!", config.getTracingSystem());
+            }
         }
+            boolean commit = !Boolean.parseBoolean(config.getEnableAutoCommit());
+            KafkaConsumer consumer = new KafkaConsumer(props);
+            consumer.subscribe(Collections.singletonList(config.getTopic()));
```
Comment (ppatierno) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: wrong indentation

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 24]
```diff
@@ -30,36 +29,43 @@ public static void main(String[] args) {
...
-            for (ConsumerRecord<String, String> record : records) {
-                log.info("Received message:");
-                log.info("\tpartition: {}", record.partition());
-                log.info("\toffset: {}", record.offset());
-                log.info("\tvalue: {}", record.value());
-                if (record.headers() != null) {
-                    log.info("\theaders: ");
-                    for (Header header : record.headers()) {
-                        log.info("\t\tkey: {}, value: {}", header.key(), new String(header.value()));
+            while (receivedMsgs < config.getMessageCount()) {
```
Comment (ppatierno) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: wrong indentation

**Review by ppatierno — CHANGES_REQUESTED** [review]

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 25]
```diff
@@ -30,13 +29,19 @@ public static void main(String[] args) {
...
 
-            props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentelemetry.instrumentation.kafkaclients.TracingConsumerInterceptor.class.getName());
+            } else {
+                log.error("Error: TRACING_SYSTEM {} is not recognized or supported!", config.getTracingSystem());
+            }
         }
-
```
Comment (kyguy) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java:39]: noise

File: java/kafka/producer/src/main/java/KafkaProducerExample.java (hunk 20 above)
Comment (kyguy) [reply java/kafka/producer/src/main/java/KafkaProducerExample.java:41]: This line still needs to be removed

File: java/kafka/producer/src/main/java/KafkaProducerExample.java (hunk 10 above)
Comment (kyguy) [reply java/kafka/producer/src/main/java/KafkaProducerExample.java:42]: It looks like the initialization line got lost in the shuffle of revisions. We need to add:
```
AutoConfiguredOpenTelemetrySdk.initialize();
```
Here and in the consumer code as well

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 26]
```diff
@@ -30,13 +29,19 @@ public static void main(String[] args) {
...
+        TracingSystem tracingSystem = config.getTracingSystem();
+        if (tracingSystem != TracingSystem.NONE) {
+            if (tracingSystem == TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
 
-            props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentelemetry.instrumentation.kafkaclients.TracingConsumerInterceptor.class.getName());
```
Comment (kyguy) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java:38]: It looks like the initialization line got lost in the shuffle of revisions. We need to add:
```
AutoConfiguredOpenTelemetrySdk.initialize();
```
Here and in the consumer code as well

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks for this work!

File: java/kafka/common/src/main/java/TracingSystem.java [hunk 27]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+public enum TracingSystem {
```
Comment (scholzj) [inline java/kafka/common/src/main/java/TracingSystem.java:6]: This seems like a huge overkill to have a module just for a single enum. I would probably integrate it into the modules directly. (But maybe it should have a further discussion since my guess is that this is how it should be done :-o).

Are there at least any plans to add more stuff to it in the next PRs?

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 28]
```diff
@@ -30,13 +30,20 @@ public static void main(String[] args) {
...
         int receivedMsgs = 0;
 
-        if (System.getenv("JAEGER_SERVICE_NAME") != null)   {
-            Tracer tracer = Configuration.fromEnv().getTracer();
-            GlobalTracer.registerIfAbsent(tracer);
+        TracingSystem tracingSystem = config.getTracingSystem();
+        if (tracingSystem != TracingSystem.NONE) {
+            if (tracingSystem == TracingSystem.JAEGER) {
+                Tracer tracer = Configuration.fromEnv().getTracer();
+                GlobalTracer.registerIfAbsent(tracer);
```
Comment (scholzj) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: If you already have the `common` module, the tracer initialization can probably go there.

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 29]
```diff
@@ -30,13 +30,20 @@ public static void main(String[] args) {
...
 
-            props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentracing.contrib.kafka.TracingConsumerInterceptor.class.getName());
+            } else if (tracingSystem == TracingSystem.OPENTELEMETRY) {
+                AutoConfiguredOpenTelemetrySdk.initialize();
```
Comment (scholzj) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java]: Same as above, if you already have the `common` module and want to keep it, the SDK initilization should probably go there.

File: java/kafka/consumer/src/main/java/KafkaConsumerExample.java [hunk 30]
```diff
@@ -30,13 +30,20 @@ public static void main(String[] args) {
...
+                props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, io.opentelemetry.instrumentation.kafkaclients.TracingConsumerInterceptor.class.getName());
+            } else {
+                log.error("Error: TRACING_SYSTEM {} is not recognized or supported!", config.getTracingSystem());
+            }
         }
-
```
Comment (scholzj) [inline java/kafka/consumer/src/main/java/KafkaConsumerExample.java:39]: Feel free to ignore it ... but keepeing and empty line after a long `if` does not really hurt. It usually makes the code readable.

File: java/kafka/deployment-tracing.yaml [hunk 31]
```diff
@@ -56,6 +56,16 @@ spec:
...
+          - name: OTEL_SERVICE_NAME
+            value: kafka-otel
+          - name: OTEL_TRACES_EXPORTER
+            value: jaeger
+          - name: OTEL_METRICS_EXPORTER
+            value: none
+          - name: OTEL_EXPORTER_JAEGER_ENDPOINT
+            value: http://my-jaeger-collector:14250
+          - name: TRACING_SYSTEM
+            value: "opentelemetry"
```
Comment (scholzj) [inline java/kafka/deployment-tracing.yaml]: Here and in all the other files =>
* How does it work together when you mix the Jaeger and OTEL environment variables?
* Maybe you should also keep the tracing example file for Jaeger as it was (and perhaps rename it to `deployment-tracing-jaeger.yaml`) and add new file for OTel tracing.

**Review by scholzj — COMMENTED** [review]
Pro-tip: If you resolve the addressed comments, the PR will be much more readable.

**Review by scholzj — APPROVED** [review]
LGTM. Thanks for the PR.

@ppatierno @mimaison Do you need another pass through it? Or can it be merged?

**Comment by ppatierno** [comment]
It's fine with me
